llvm/lib/Target/AIE/AIEWawRegRewriter.cpp

```diff
@@ -0,0 +1,435 @@
+
+  bool isLastDef(const MachineOperand &Def, const MachineInstr &DefInstr,
+                 const MachineBasicBlock &MBB);
+
+  bool reuseSubRegs(SmallVector<const MachineInstr *, 4> &OrigInstr,
+                    SmallVector<const MachineInstr *, 4> &OtherInstr);
+
+  bool containEqualSubRegDefs(const MachineInstr *InstrA,
+                              const MachineInstr *InstrB);
+
+  void addAliasRegs(SmallSet<MCPhysReg, 16> &ForbiddenPhysRegs,
+                    const MCPhysReg &PhysReg);
+
+  Register getPhysReg(const MachineOperand &MO) const;
+};
+
+Register AIEWawRegRewriter::getPhysReg(const MachineOperand &MO) const {
+  Register PhysReg = MO.getReg();
+  if (PhysReg.isVirtual())
+    PhysReg = VRM->getPhys(PhysReg);
+  return PhysReg;
+}
+
+bool AIEWawRegRewriter::runOnMachineFunction(MachineFunction &MF) {
+
+  MachineLoopInfo &MLI = getAnalysis<MachineLoopInfo>();
+
+  SmallVector<const MachineBasicBlock *, 4> LoopMBBs =
+      AIELoopUtils::getInnermostLoopBlocks(MLI);
+
+  if (LoopMBBs.empty())
+    return false;
+
+  this->MF = &MF;
+  MRI = &MF.getRegInfo();
+  TRI = static_cast<const AIEBaseRegisterInfo *>(MRI->getTargetRegisterInfo());
+  VRM = &getAnalysis<VirtRegMap>();
+  LRM = &getAnalysis<LiveRegMatrix>();
+  LIS = &getAnalysis<LiveIntervals>();
+
+  LLVM_DEBUG(dbgs() << "*** WAW Loop Register Rewriting: " << MF.getName());
+  LLVM_DEBUG(dbgs() << " ***\n");
+
+  return renameMBBPhysRegs(LoopMBBs);
+}
+
+bool AIEWawRegRewriter::renameMBBPhysRegs(
+    const SmallVector<const MachineBasicBlock *, 4> &LoopMBBs) {
+  bool Modified = false;
+  for (const MachineBasicBlock *MBB : LoopMBBs) {
+    LLVM_DEBUG(dbgs() << "WAW Reg Renaming BasicBlock "; MBB->dump();
+               dbgs() << "\n");
+    SmallVector<MCPhysReg, 16> UsedPhyRegs;
+    SmallVector<MCPhysReg, 16> UsedVirtRegs;
+    SmallVector<std::pair<MCRegister, MCRegister>, 8> ReplacementRegs;
+
+    SmallSet<MCPhysReg, 16> ForbiddenPhysRegs =
+        getMBBForbiddenReplacements(MBB);
+
+    std::map<unsigned, SmallVector<const MachineInstr *, 4>> UndefPartReuse;
+
+    for (const MachineInstr &MI : *MBB) {
+
+      if (isIdentityCopy(MI))
+        continue;
+
+      for (const MachineOperand &Op : MI.defs()) {
+        Register Reg = Op.getReg();
+        if (!Reg.isVirtual())
+          continue;
+        if (VRM->hasRequiredPhys(Reg))
+          continue;
+        if (Op.isTied())
+          continue;
+        if (!TRI->isVecOrAccRegClass(MRI->getRegClass(Reg)))
+          continue;
+        //  Subregisters can reuse Constants in the high part of the
+        //  register, therefore renaming a reused physical register can
+        //  prohibit the elimination of a move of constants, thus degrading
+        //  performance if this physical register is removed
+        if (!DisableCopySubRegHandling &&
+            isUndefReserved(Op, MI, *MBB, UndefPartReuse))
+          continue;
```

Comment:
I wonder if it wouldn't just be enough to prevent renaming registers that are defined through sub-registers? Or maybe make the check more generic and preventing renaming when one lane is defined through a COPY. (Because then we might prevent coalescing)

Reply:
If a sub register has a copy operation, it is immediatly ignored for renaming